SDI-1659: support for CORS header

candysmurf opened the pull request:
Fixes #998 

Summary of changes:
- Added CORS support.
- Cors origins have to be specified in the global config. Otherwise it's the same origin.

Testing done:
- swagger-ur which is in a different origin of Snap.

@intelsdi-x/snap-maintainers

andrzej-k commented on mgmt/rest/server.go:
```diff
@@ -92,6 +93,15 @@ func New(cfg *Config) (*Server, error) {
+
+	// Deals with CORS
+	c := cors.New(cors.Options{
+		AllowedOrigins: []string{"*"},
```
I don't think that it's safe to allow unrestricted access from any domain. Is there a possibility to constrain allowed origins - via snapteld configuration maybe?

candysmurf replied:
@andrzej-k, agreed your concerns. Should we allow localhost for now until we know whoch domains are allowed?

jcooklin replied:
Configurable with 'localhost' as the default?

candysmurf replied:
@jcooklin, sounds good. I'll add it as a global option.

marcintao commented on mgmt/rest/server.go:
```diff
@@ -92,6 +93,15 @@ func New(cfg *Config) (*Server, error) {
+	// Deals with CORS
+	c := cors.New(cors.Options{
+		AllowedOrigins: []string{"*"},
+		AllowedMethods: []string{"GET, POST, DELETE, PUT, PATCH, OPTIONS"},
```
Not all of the HTTP methods were allowed previously. I believe it would be the safest to keep the list of allowed methods to a required minimum.
`PATCH` - I didn't find it declared for any of REST controllers. Are you sure we need to allow it? 
`OPTIONS` - the same as for `PATCH`. It's not needed for pre-flight requests. Negroni allows it implicitly only for the purpose of CORS. Furthermore, if it's needed for Snap API, it should be enabled with a flag `OptionsPassthrough`.

candysmurf replied:
OPTIONS is needed for preflight. I can remove PATCH for now.

marcintao replied:
Yes, you're right - OPTIONS are needed for preflight. But isn't it handled before the actual request? I'm thinking about this: https://github.com/rs/cors/blob/master/cors.go#L189. I guess the actual request handling and method checking takes place later.

candysmurf replied:
@marcintao, thanks for looking into it. However, Snap has two places to be patched. One place cannot use the lib. I added maxAge to ease your concerns. Please let me know any other suggestions you have.

candysmurf commented:
@andrzej-k, @marcintao, @jcooklin, CORS PR is update. Please review when you have a chance.

andrzej-k commented on mgmt/rest/server_test.go:
```diff
@@ -161,5 +163,45 @@ func TestRestAPIDefaultConfig(t *testing.T) {
+		Convey("Origins match results", func() {
+			So(s.allowedOrigins, ShouldResemble, map[string]bool{"http://127.0.0.1:80": true, "http://example.com": true})
+			So(len(o), ShouldEqual, 2)
+		})
```
Consider adding more rainy day scenarios, ie. incorrect origin format, incorrect separator, etc.

andrzej-k commented on mgmt/rest/server.go:
```diff
@@ -259,6 +304,18 @@ func (s *Server) addRoutes() {
+	for _, o := range os {
+		s.allowedOrigins[strings.TrimSpace(o)] = true
+	}
+	return os
```
Consider adding more validation, for example if origin format is correct, something like: 
- httpa://domain.com
- 127.0.0.0.1
- http://?domain.com
- ...

should not pass, and user should be warned early that configuration is not valid.

candysmurf commented on mgmt/rest/server.go:
```diff
@@ -60,6 +70,7 @@ type Server struct {
 	// the following instance variables are used to cleanly shutdown the server
 	serverListener net.Listener
 	closingChan    chan bool
+	allowedOrigins map[string]bool
```
not sure your comment here. It's used in authMiddleware method.

andrzej-k replied:
I'm just saying that "allowedOrigins" variable is added in a section which is described with a comment (3 lines above): 
```// the following instance variables are used to cleanly shutdown the server```

This may be misleading, so I just proposed to move "allowedOrigins" before that comment, i.e. 4 lines up.

candysmurf commented:
@andrzej-k, thanks for your awesome review. I changed the name, added URL validation and more negative tests as you suggested. Please let me know if anything else.

candysmurf commented:
good catch! Changed. Thanks @andrzej-k!

jcooklin commented:
LGTM
